src/envvars.zig

```diff
@@ -0,0 +1,311 @@
+    pub const internal_suppress_crash_in_bun_run = newFeatureFlag("BUN_INTERNAL_SUPPRESS_CRASH_IN_BUN_RUN");
+    pub const internal_suppress_crash_on_napi_abort = newFeatureFlag("BUN_INTERNAL_SUPPRESS_CRASH_ON_NAPI_ABORT");
+    pub const internal_suppress_crash_on_process_kill_self = newFeatureFlag("BUN_INTERNAL_SUPPRESS_CRASH_ON_PROCESS_KILL_SELF");
+    pub const internal_suppress_crash_on_uv_stub = newFeatureFlag("BUN_INTERNAL_SUPPRESS_CRASH_ON_UV_STUB");
+    pub const last_modified_pretend_304 = newFeatureFlag("BUN_FEATURE_FLAG_LAST_MODIFIED_PRETEND_304");
+    pub const no_codesign_macho_binary = newFeatureFlag("BUN_NO_CODESIGN_MACHO_BINARY");
+    pub const no_libdeflate = newFeatureFlag("BUN_FEATURE_FLAG_NO_LIBDEFLATE");
+    pub const node_no_warnings = newFeatureFlag("NODE_NO_WARNINGS");
+    pub const trace = newFeatureFlag("BUN_TRACE");
+};
+
+const EnvVarType = enum { string, boolean };
+
+/// Create a new envionment variable definition.
+///
+/// The resulting type has methods for interacting with the environment variable.
+///
+/// Technically, none of the operations here are thread-safe, so writing to environment variables
+/// does not guarantee that other threads will see the changes. You should avoid writing to
+/// environment variables.
+fn new(comptime T: EnvVarType, comptime key: [:0]const u8, comptime opts: EnvVarOpts(T)) type {
+    return platformSpecificNew(T, key, key, opts);
+}
+
+/// Identical to new, except it allows you to specify different keys for POSIX and Windows.
+///
+/// If the current platform does not have a key specified, all methods that attempt to read the
+/// environment variable will fail at compile time, except for `platformGet` and `platformKey`,
+/// which will return null instead.
+fn platformSpecificNew(
+    comptime T: EnvVarType,
+    comptime posix_key: ?[:0]const u8,
+    comptime windows_key: ?[:0]const u8,
+    comptime opts: EnvVarOpts(T),
+) type {
+    const comptime_key: []const u8 =
+        if (posix_key) |pk| pk else if (windows_key) |wk| wk else "<undefined>";
+
+    if (posix_key == null and windows_key == null) {
+        @compileError("Environment variable " ++ comptime_key ++ " has no keys for POSIX " ++
+            "nor Windows specified. Provide a key for either POSIX or Windows.");
+    }
+
+    const KeyType = [:0]const u8;
+    // Represents the return type of public methods, ignoring the optionality.
+    const NonOptionalReturnType = comptime switch (T) {
+        .string => []const u8,
+        .boolean => bool,
+    };
+
+    // The actual return type of public methods.
+    const ReturnType = if (opts.default != null) NonOptionalReturnType else ?NonOptionalReturnType;
+
+    // Type used to communicate between the Cache and the public interface of this env var.
+    const CachedInterface = union(enum) {
+        /// The environment variable hasn't been loaded yet.
+        undefined: void,
```

Comment:
imo we should not call this 'undefined' because that has another meaning in zig. it could be 'unloaded' instead

Reply:
yes, `not_cached` or `unknown` would also work

Reply:
How about `unset`, I recall having seen that across quite a lot of codebases

Reply:
that would be confusing with `not_set`